File: packages/evals/tasks/bench/extract/extract_geniusee_2.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";

export default defineBenchTask(
  { name: "extract_geniusee_2" },
  async ({ logger, debugUrl, sessionUrl, stagehand, page }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/geniusee/");
      const selector = "xpath=/html/body/main/div[2]/div[2]/div[2]/table/tbody/tr[9]";
      const { data: scalability } = await stagehand.extract(
        "Extract the scalability comment in the table for Gemini (Google)",
        z.object({
          scalability: z.string(),
        }),
        { selector: selector },
      );

      const scalabilityComment = scalability.scalability;

      // scalabilityCommentWeShouldNotGet matches a scalability comment in the table,
      // but since we are using targeted_extract here,
      // and passing in a selector that does NOT contain the scalabilityCommentWeShouldNotGet,
      // the LLM should have no visibility into scalabilityCommentWeShouldNotGet if
      // targeted_extract is performing correctly
      const scalabilityCommentWeShouldNotGet = {
        scalability: "Scalable architecture with API access",
      };

      const commentMatches = scalabilityComment == scalabilityCommentWeShouldNotGet.scalability;

      if (commentMatches) {
        logger.error({
          message:
            "extracted scalability comment matches the scalability comment that we SHOULD NOT get",
          level: 0,
          auxiliary: {
            expected: {
              value: scalabilityCommentWeShouldNotGet.scalability,
              type: "string",
            },
            actual: {
              value: scalabilityComment,
              type: "string",
            },
          },
        });
        return {
          _success: false,
          error: "scalability comment matches the scalability comment that we SHOULD NOT get",
          logs: logger.getLogs(),
          debugUrl,
          sessionUrl,
        };
      }

      return {
        _success: true,
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    }
  },
);

File: packages/sdk-ruby/evals/tasks/extract/extract_geniusee_2.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/extract/extract_geniusee_2.ts

Evals.define_task("extract_geniusee_2") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/geniusee/")
  # The locator engine prefix is required for XPath selectors.
  locator = t.page.locator("xpath=/html/body/main/div[2]/div[2]/div[2]/table/tbody/tr[9]")
  result = t.stagehand.extract(
    "Extract the scalability comment in the table for Gemini (Google)",
    schema: {
      "type" => "object",
      "properties" => { "scalability" => { "type" => "string" } },
      "required" => ["scalability"],
      "additionalProperties" => false,
    },
    locator: locator,
  )

  scalability_comment = result.data["scalability"]

  # scalability_comment_we_should_not_get matches a scalability comment in the table,
  # but since we are using targeted_extract here,
  # and passing in a selector that does NOT contain the scalability_comment_we_should_not_get,
  # the LLM should have no visibility into scalability_comment_we_should_not_get if
  # targeted_extract is performing correctly
  scalability_comment_we_should_not_get = "Scalable architecture with API access"

  if scalability_comment == scalability_comment_we_should_not_get
    t.logger.error("extracted scalability comment matches the scalability comment that we SHOULD NOT get",
                   { expected: scalability_comment_we_should_not_get, actual: scalability_comment })
    next {
      _success: false,
      error: "scalability comment matches the scalability comment that we SHOULD NOT get",
    }
  end

  { _success: true, scalabilityComment: scalability_comment }
end
